### PR TITLE
feat(cmd): flipt evaluate CLI subcommand

### LINKED ISSUE #2254: [FLI-636] `flipt evaluate` CLI subcommand
### Problem

It would be nice to have a quick CLI command for running one (or more) evaluations against the evaluation API.
This would be handy for demoing Flipt capabilities, checking connectivity, experimenting with flag configuration or even, with added telemetry, understanding client and server side latency.

### Ideal Solution

A subcommand on the flipt CLI, which makes an evaluation, with support for all the evaluation parameters.

```console
# perhaps entity is random for each eval unless specified via -e
flipt evaluate -e <entity-id> [... key=value] # where key=value are context key/value pairs

# alternative using a more JSON like syntax for context
flipt evaluate {"key":"value"}
```

Ideally, the response would contain all the evaluate response details.
However, (perhaps behind some verbose flag) it would be nice to get some telemetry or additional response metadata out too. For example, it would be nice to get some client and server side latencies.

Bonus: ability to set it into a polling or watch mode. Mostly so that you can see latency over time, or adjust flag state and watch how it changes live.

e.g. `flipt evaluate -w ...` (for watch)

### Search

- [X] I searched for other open and closed issues before opening this

### Additional Context

_No response_

<sub>[FLI-636](https://linear.app/flipt/issue/FLI-636/flipt-evaluate-cli-subcommand)</sub>

### PR DESCRIPTION
**Pull request by erka**
fixes #2254

### CODE REVIEW

File: build/testing/testdata/cli.txt [hunk 1]
```diff
@@ -13,6 +13,7 @@ $ flipt --config /path/to/config.yml migrate
 Available Commands:
   bundle      Manage Flipt bundles
   config      Manage Flipt configuration
+  evaluate    Evaluate the flag
```
Comment (markphelps) [inline build/testing/testdata/cli.txt:16]:
```suggestion
  evaluate    Evaluate a flag
```

File: cmd/flipt/evaluate.go [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+	interval      time.Duration
+	contextValues []string
+}
+
+func newEvaluateCommand() *cobra.Command {
+	c := &evaluateCommand{}
+
+	cmd := &cobra.Command{
+		Use:   "evaluate [flag]",
+		Short: "Evaluate the flag",
```
Comment (markphelps) [inline cmd/flipt/evaluate.go:33]:
```suggestion
		Short: "Evaluate a flag",
```

File: cmd/flipt/evaluate.go [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+	token         string
+	interval      time.Duration
+	contextValues []string
+}
+
+func newEvaluateCommand() *cobra.Command {
+	c := &evaluateCommand{}
+
+	cmd := &cobra.Command{
+		Use:   "evaluate [flag]",
```
Comment (markphelps) [inline cmd/flipt/evaluate.go:32]:
```suggestion
		Use:   "evaluate [flagKey]",
```

File: cmd/flipt/evaluate.go [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+		"address", "a",
+		"http://localhost:8080",
+		"address of Flipt instance to evaluate.",
+	)
+
+	cmd.Flags().StringVarP(
+		&c.token,
+		"token", "t",
+		"",
+		"client token used to authenticate access to Flipt instance when evaluating.",
```
Comment (markphelps) [inline cmd/flipt/evaluate.go:75]:
```suggestion
		"client token used to authenticate access to Flipt instance.",
```

File: cmd/flipt/evaluate.go [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+		"watch", "w",
+		false,
+		"enable watch mode.",
+	)
+
+	cmd.Flags().DurationVarP(
+		&c.interval,
+		"interval", "i",
+		time.Second,
+		"internal between requests in watch mode.",
```
Comment (markphelps) [inline cmd/flipt/evaluate.go:89]:
```suggestion
		"interval between requests in watch mode.",
```

File: cmd/flipt/evaluate.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+		"context", "c",
+		[]string{},
+		"evaluation request context as key=value.",
+	)
+
+	cmd.Flags().StringVarP(
+		&c.address,
+		"address", "a",
+		"http://localhost:8080",
+		"address of Flipt instance to evaluate.",
```
Comment (markphelps) [inline cmd/flipt/evaluate.go:68]:
```suggestion
		"address of Flipt instance.",
```

i think we might need to update some other places where we use `address` and `token` descriptions that could be a bit shorter. i can do that in another PR

**Review by markphelps — COMMENTED** [review]
looks great! thanks for taking this on @erka ! just a couple minor suggestions mostly around command descriptions/wording

**Review by markphelps — APPROVED** [review]
Works great!!

```
./bin/flipt evaluate foo --context test=bar
2024-01-20T08:14:53-05:00	INFO	variant	{"flag": "foo", "match": true, "reason": "MATCH_EVALUATION_REASON", "variantKey": "sadf", "variantAttachment": "", "segmentKeys": ["eee"], "requestId": "b2562109-6e6e-4449-82ff-aaf8aa2eda14", "requestDurationMillis": 1.269834, "timestamp": "2024-01-20T13:14:53Z"}
workspace/flipt - [cmd-evaluate] » 
```

In the future I wonder if its worth us re-thinking using the `defaultLogger` for all CLI commands and instead just writing a more simplified output (removing the `2024-01-20T08:14:53-05:00	INFO	` prefix). This would mean using different output for `flipt` server, and the CLI commands but I think it would make for a more traditional Unixy experience. wdty @GeorgeMac ?

**Comment by erka** [comment]
> ```./bin/flipt evaluate foo --context test=bar```
> 
You forgot to `--watch` it 😂

It would be great to have some spec for cli output. Currently it's a bit messy in this command.

**Comment by markphelps** [comment]
> > ```./bin/flipt evaluate foo --context test=bar```
> 
> > 
> 
> You forgot to `--watch` it 😂
> 
> 
> 
> It would be great to have some spec for cli output. Currently it's a bit messy in this command.
> 
> 
> 
> 
> 
> 

Yeah I will write up an issue and we can discuss preferred output for cli commands

**Pull request merged**